### PR TITLE
Add goto rename <from> <to>

### PR DESCRIPTION
**Pull request by Arxcis**
## Goal
- Implement the feature suggested in #87 

## Todo
- [x] Write tests
- [x] Add to help menu
- [x] Add autocompletion for rename|mv in start_goto

## Demo

[asciicast](https://asciinema.org/a/QIbyMuBlABfTk06yHQGycptkq)

```
goto list
article.as-we-may-think
bruno
code
downloads
file.goto
folder
gif
git.goto
github
github.goto
goto
home
issues
localgif
music.bruno-mars
newissue
pulls
repo
steam.downloads
steam.friends
steam.games
tesseract
toggl
universe
bash-3.2$ goto mv goto goto2
Renamed goto to goto2
bash-3.2$ goto mv goto goto2

Ah hoy!

    - The Magic word goto does not exist.

    Are you sure it exists in the project you are in now?
    Type:

        goto list

    to see all shortcuts in this project.
    
bash-3.2$ goto mv goto2 goto
Renamed goto2 to goto
bash-3.2$ goto mv goto

Ah hoy!
    Error: missing second magic word.

    Try again by adding the magicword you want to mv to:

                goto mv goto <to>
    
bash-3.2$ goto mv

Ah hoy!
    Error: missing both magicwords

    Try again by adding the magicword you want to mv
    and the magicword you want to mv to:

                goto mv <from> <to>
```

### CODE REVIEW

File: goto/commands/rename.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    to_magicword = args[1]
+
+    if from_magicword not in magic.magic:
+        return None, GotoWarning("magicword_does_not_exist", magicword=from_magicword)
+
+    if to_magicword in magic.magic:
+        return None, GotoWarning("adding_existing_magicword_short", magicword=to_magicword, uri=magic[to_uri])
+
+    from_uri = magic[from_magicword]
```
Comment (technocake) [inline goto/commands/rename.py]: I think the handling of the renaming should be done inside the magic class. 
This is in order to keep a consitent separation of concern - and thererfore the possibilty to interchange the underlaying implementation.

The rename operation should therefore be moved to a method in the GotoMagic class. `rename` or something

File: goto/gotomagic/text/warning.py [hunk 2]
```diff
@@ -79,6 +87,25 @@
                 goto {command} <magicword>
     """,
 
+    missing_both_magicwords="""
+Ah hoy!
```
Comment (technocake) [inline goto/gotomagic/text/warning.py]: These warning messages are great :)!

File: goto/the_real_goto.py [hunk 3]
```diff
@@ -76,6 +76,9 @@ def run_command(magic, command, args):
     if command == 'cd':
         return commands.cd(magic, args)
 
+    if command in ['mv', 'rename']:
```
Comment (technocake) [inline goto/the_real_goto.py]: Elegant

File: goto/commands/rename.py (hunk 1 above)
Comment (Arxcis) [reply goto/commands/rename.py]: I totally agree

**Comment by technocake** [comment]
Tests are written in #95

**Comment by technocake** [comment]
@Arxcis please review the latest two commits:
 

- **Autocompletions for rename** https://github.com/technocake/goto/pull/89/commits/9fc1d79bd33edb6da02f31f75f1a80e2e1b8867e
- **Update help text** https://github.com/technocake/goto/pull/89/commits/9b44bad096affed53f7883d8a0d8de499f0fdf7a

File: .gitignore [hunk 4]
```diff
@@ -3,6 +3,9 @@
 # active project is stored in this file
 active-project
 
+*.sublime-workspace
```
Comment (technocake) [inline .gitignore]:
```
Projects in Sublime Text are made up of two files: the .sublime-project file, which contains the project definition, and the .sublime-workspace file, which contains user specific data, such as the open files and the modifications to each. 
```

https://www.sublimetext.com/docs/3/projects.html

Comment (Arxcis) [reply .gitignore]: goto workspace?

File: README.md [hunk 5]
```diff
@@ -49,28 +49,29 @@ your rcfile should do the trick in most cases.
...
-    
-...this command will open folder with Sublime Text
-    goto subl
-
-...IntelliJ:  goto idea,  VSCode: goto vscode                               
+    The basics
+        goto <magicword>                        Go to shortcut
+        goto add    <magicword> <url or path>   Add shortcut
+        goto update <magicword> <new url/path>  Update shortcut
+        goto rename <magicword> <new name>      Rename shortcut
```
Comment (Arxcis) [inline README.md]: Good

File: goto/commands/usage.py [hunk 6]
```diff
@@ -12,7 +12,8 @@ def usage():
     The basics
         goto <magicword>                        Go to shortcut
         goto add    <magicword> <url or path>   Add shortcut
-        goto update <magicword> <url or path>   Update shortcut
+        goto update <magicword> <new url/path>  Update shortcut
+        goto rename <magicword> <new name>      Rename shortcut
```
Comment (Arxcis) [inline goto/commands/usage.py]: good

File: goto/gotomagic/magic.py [hunk 7]
```diff
@@ -61,6 +62,19 @@ def update_shortcut(self, magicword, uri):
             )
             exit(1)
 
+    def rename_shortcut(self, from_magicword, to_magicword, overwrite=False):
+        """ Renaming a shortcut """
+        if not overwrite and to_magicword in self.magic:
+            uri = self.magic[to_magicword]
+            return GotoWarning('adding_existing_magicword_short', magicword=to_magicword, uri=uri)  # noqa
```
Comment (Arxcis) [inline goto/gotomagic/magic.py]: What does #noqa mean?

File: goto/gotomagic/magic.py [hunk 8]
```diff
@@ -61,6 +62,19 @@ def update_shortcut(self, magicword, uri):
...
+
+        if from_magicword not in self.magic:
+            return GotoWarning('magicword_does_not_exist', magicword=from_magicword)  # noqa
+
+        from_uri = self.magic[from_magicword]
```
Comment (Arxcis) [inline goto/gotomagic/magic.py]: uri or url? Should we just go for one of them? I vote for url 👍 Goto is a locator tool. It gives you locations of stuff, based on uniform resource locators.

File: .gitignore (hunk 4 above)
Comment (technocake) [reply .gitignore]: yeah, I added a sublime project for goto, in order to make the sublime Goto (cmd + P) not list entries in the .tox folder.  It annoyed me that jumping to different files involved scrolling down on the list to select the right source file. 

Ofcourse sublime then adds a workspace file which is highly personal and only relevant to the current sublime session. It is recommended to not check in to source control.

File: goto/gotomagic/magic.py (hunk 7 above)
Comment (technocake) [reply goto/gotomagic/magic.py]: no quality assurrance. It effictively turns off pep8 linting on that mine. making that line be an explicit exemption of the pep8 rules. the line is too long. thats why. 

If you enable pep8 linting in your editor, you get warnings when the code style breaks pep8 style.

File: goto/gotomagic/magic.py (hunk 8 above)
Comment (technocake) [reply goto/gotomagic/magic.py]: We aee actually working with uris, and some of the uris we are working with are urls.

I actually would want developers going deep enough to contribute to goto, to be intrigued by the use of uri in sthe source code and ask themself, what is the difference between uris and urls.


In the help text however, I think it is more.. practical, to use urls over uris. to ease adoptation of new users. 

the file paths for instance, are not urls, but they are uris.

Comment (Arxcis) [reply goto/gotomagic/magic.py]: Meanwhile
```
jonas@8600k:~$ project add uri-vs-url
active project is now: urivsurl

jonas@8600k:~$ goto list

jonas@8600k:~$ goto add venn-diagram-1 https://res.cloudinary.com/practicaldev/image/fetch/s--lrbx3qNQ--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://thepracticaldev.s3.amazonaws.com/i/j4bka41nypm4do1f3e5b.JPG
Added magic word venn-diagram-1

jonas@8600k:~$ goto add what-is-the-difference-between-a-uri-and-a-url https://dev.to/flippedcoding/what-is-the-difference-between-a-uri-and-a-url-4455
Added magic word what-is-the-difference-between-a-uri-and-a-url

jonas@8600k:~$ goto add venn-diagram-2 https://danielmiessler.com/images/uri-urn-url-miessler-2019.png
Added magic word venn-diagram-2

jonas@8600k:~$ goto add the-difference-between-urls-uris-and-urns https://danielmiessler.com/study/url-uri/
Added magic word the-difference-between-urls-uris-and-urns

jonas@8600k:~$ goto list
the-difference-between-urls-uris-and-urns
venn-diagram-1
venn-diagram-2
what-is-the-difference-between-a-uri-and-a-url

jonas@8600k:~$ 
```

Comment (Arxcis) [reply goto/gotomagic/magic.py]: rfc3305#confusion - https://tools.ietf.org/html/rfc3305#section-2.3

**Pull request merged**